driver-scala/src/main/scala/org/mongodb/scala/internal/FlatMapObservable.scala

```diff
@@ -18,72 +18,99 @@ package org.mongodb.scala.internal
-
-              def request(n: Long): Unit = {
+              override def unsubscribe(): Unit = subscription.unsubscribe()
+              override def request(n: Long): Unit = {
                 require(n > 0L, s"Number requested must be greater than zero: $n")
                 val localDemand = addDemand(n)
-                val (sub, num) = nestedSubscription.map((_, localDemand)).getOrElse((subscription, 1L))
-                sub.request(num)
+                state.get() match {
+                  case Init              => subscription.request(1L)
+                  case WaitingOnChild(s) => s.request(localDemand)
+                  case _                 => // noop
+                }
               }
-
-              override def unsubscribe(): Unit = subscription.unsubscribe()
             }
-
             outerSubscription = Some(masterSub)
+            state.set(Init)
             observer.onSubscribe(masterSub)
           }
 
           override def onComplete(): Unit = {
-            if (!onCompleteCalled) {
-              onCompleteCalled = true
-              if (nestedSubscription.isEmpty) observer.onComplete()
+            state.get() match {
+              case Done  => // ok
+              case Error => // ok
+              case Init if state.compareAndSet(Init, Done) =>
+                observer.onComplete()
+              case w @ WaitingOnChild(_) if state.compareAndSet(w, LastChildNotified) =>
+              // letting the child know that we delegate onComplete call to it
+              case LastChildNotified =>
+              // wait for the child to do the delegated onCompleteCall
+              case LastChildResponded if state.compareAndSet(LastChildResponded, Done) =>
+                observer.onComplete()
+              case other =>
+                // state machine is broken, let's fail
+                // normally this won't happen
+                throw new IllegalStateException(s"Unexpected state in FlatMapObservable `onComplete` handler: ${other}")
             }
           }
 
-          override def onError(throwable: Throwable): Unit = observer.onError(throwable)
+          override def onError(throwable: Throwable): Unit = {
+            observer.onError(throwable)
+          }
 
           override def onNext(tResult: T): Unit = {
             f(tResult).subscribe(
               new Observer[S]() {
                 override def onError(throwable: Throwable): Unit = {
-                  nestedSubscription = None
+                  state.set(Error)
                   observer.onError(throwable)
                 }
 
                 override def onSubscribe(subscription: Subscription): Unit = {
-                  nestedSubscription = Some(subscription)
+                  state.set(WaitingOnChild(subscription))
                   if (demand > 0) subscription.request(demand)
                 }
 
                 override def onComplete(): Unit = {
-                  nestedSubscription = None
```

Comment:
for the sake of completeness, race condition (threads A & B):

A: `Observer[S]#onComplete`
A: `nestedSubscription = None`
B: `Observer[T]#onComplete`
B: `onCompleteCalled == false` => B: `onCompleteCalled = true` 
B: `nestedSubscription.isEmpty == true` => B: `observer.onComplete()`
A: `onCompleteCalled == true` => A: `observer.onComplete`

At least this is an explanation I've come up with while trying to debug the failing test (under original code)